### PR TITLE
Add new ArrayBuilder & Improve merge result

### LINKED ISSUE #3496: improve transform `merge_result` temp datavalue
**Summary**

Now in pipelines transform, when aggregating final results, especially in `group by` function, there may be many temp `datavalues`, as code below `Vec<Vec<DataValue>>`.

Consider remove these temp values and just write value into arrow array(maybe binary array, primitive array...)

https://github.com/datafuselabs/databend/blob/610dba09b15aa6dac3f2e0355c085b446d4e5e74/query/src/pipelines/transforms/transform_group_by_final.rs#L181-L198

### PR DESCRIPTION
**Pull request by Veeupup**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

* Add new `MutableArrayBuilder`s
* Improve `merge_result` in `aggregate` function and `pipeline`

### TODO

* make `MutableArrayBuilder` push `val` to `MutableBuffer` directly to avoid `try_push` branch prediction failure.

## Changelog

- Improvement

## Related Issues

Fixes #3496 

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

File: common/functions/src/aggregates/aggregate_arg_min_max.rs [hunk 1]
```diff
@@ -163,8 +165,128 @@ where
         Ok(())
     }
 
-    fn merge_result(&mut self) -> Result<DataValue> {
-        Ok(self.data.clone())
+    #[allow(unused_mut)]
+    fn merge_result(&mut self, array: &mut dyn MutableArray) -> Result<()> {
+        // TODO(veeupup) refactor them to use macro
+        match self.data {
+            DataValue::Int8(val) => {
```
Comment (sundy-li) [inline common/functions/src/aggregates/aggregate_arg_min_max.rs:172]: MutablePrimitiveArray::<T>

File: common/functions/src/aggregates/aggregate_avg.rs [hunk 2]
```diff
@@ -147,14 +150,136 @@ where
...
-            return Ok(DataValue::Float64(None));
+            array.push_null();
+            return Ok(());
         }
+
         let v: f64 = NumCast::from(state.value).unwrap_or_default();
-        Ok(DataValue::Float64(Some(v / state.count as f64)))
+        let val = v / state.count as f64;
+
+        match array.data_type() {
```
Comment (sundy-li) [inline common/functions/src/aggregates/aggregate_avg.rs:165]: f64 is enough.

Comment (Veeupup) [reply common/functions/src/aggregates/aggregate_avg.rs:165]: resolved

File: common/functions/src/aggregates/aggregate_arg_min_max.rs [hunk 3]
```diff
@@ -163,8 +166,51 @@ where
         Ok(())
     }
 
-    fn merge_result(&mut self) -> Result<DataValue> {
-        Ok(self.data.clone())
+    #[allow(unused_mut)]
+    fn merge_result(&mut self, array: &mut dyn MutableArray) -> Result<()> {
```
Comment (Veeupup) [inline common/functions/src/aggregates/aggregate_arg_min_max.rs:170]: boolean datatype

File: common/functions/src/aggregates/aggregate_min_max.rs [hunk 4]
```diff
@@ -134,8 +134,37 @@ where
         Ok(())
     }
 
-    fn merge_result(&mut self) -> Result<DataValue> {
-        Ok(self.value.into())
+    #[allow(unused_mut)]
+    fn merge_result(&mut self, array: &mut dyn MutableArrayBuilder) -> Result<()> {
```
Comment (sundy-li) [inline common/functions/src/aggregates/aggregate_min_max.rs:138]: No need to match.

File: common/functions/src/aggregates/aggregate_sum.rs [hunk 5]
```diff
@@ -163,9 +163,38 @@ where
         Ok(())
     }
 
-    fn merge_result(&self, place: StateAddr) -> Result<DataValue> {
+    #[allow(unused_mut)]
+    fn merge_result(&self, place: StateAddr, array: &mut dyn MutableArrayBuilder) -> Result<()> {
         let state = place.get::<AggregateSumState<SumT>>();
```
Comment (sundy-li) [inline common/functions/src/aggregates/aggregate_sum.rs:168]: No need to match.

File: common/datavalues/src/arrays/primitive/mutable.rs [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn as_mut_any(&mut self) -> &mut dyn std::any::Any {
+        self
+    }
+
+    fn as_arc(&mut self) -> Arc<dyn Array> {
```
Comment (sundy-li) [inline common/datavalues/src/arrays/primitive/mutable.rs:49]: transform to as_series.

File: query/src/pipelines/transforms/transform_aggregator_final.rs [hunk 7]
```diff
@@ -121,19 +125,50 @@ impl Processor for AggregatorFinalTransform {
...
-            let merge_result = func.merge_result(place)?;
-            final_result.push(merge_result.to_series_with_size(1)?);
+            let array: &mut dyn MutableArrayBuilder = aggr_values[idx].borrow_mut();
+            let _ = func.merge_result(place, array)?;
         }
 
+        let mut columns: Vec<Series> = Vec::with_capacity(funcs_len);
+        for mut array in aggr_values {
+            let datatype = array.data_type();
+            with_match_primitive_type!(datatype, |$T| {
```
Comment (sundy-li) [inline query/src/pipelines/transforms/transform_aggregator_final.rs:148]: array.as_series()

File: query/src/pipelines/transforms/transform_group_by_final.rs [hunk 8]
```diff
@@ -179,33 +180,56 @@ impl Processor for GroupByFinalTransform {
...
 
                 // Build final state block.
                 let mut columns: Vec<Series> = Vec::with_capacity(aggr_funcs_len + group_expr_len);
-
-                for (i, value) in aggr_values.iter().enumerate() {
-                    columns.push(DataValue::try_into_data_array(
-                        value.as_slice(),
-                        &self.aggr_exprs[i].to_data_type(&self.schema_before_group_by)?,
-                    )?);
+                for mut array in aggr_values {
```
Comment (sundy-li) [inline query/src/pipelines/transforms/transform_group_by_final.rs:206]: as_series

File: common/functions/src/aggregates/aggregate_min_max.rs (hunk 4 above)
Comment (Veeupup) [reply common/functions/src/aggregates/aggregate_min_max.rs:138]: resolved

**Review by sundy-li — APPROVED** [review]

**Review by zhang2014 — APPROVED** [review]

**Comment by BohuTANG** [comment]
ci failure:
LLVM ERROR: IO failure on output stream: No space left on device
923
error: could not compile `databend-query`
924
Error: The process 'cross' failed with exit code 101

**Comment by BohuTANG** [comment]
@mergify update

**Comment by Veeupup** [comment]
> ci failure: LLVM ERROR: IO failure on output stream: No space left on device 923 error: could not compile `databend-query` 924 Error: The process 'cross' failed with exit code 101



> ci failure: LLVM ERROR: IO failure on output stream: No space left on device 923 error: could not compile `databend-query` 924 Error: The process 'cross' failed with exit code 101

it seems today's CI all failed...

**Comment by BohuTANG** [comment]
@mergify update

**Comment by Veeupup** [comment]
@mergify update

**Pull request merged**